diff --git a/Makefile b/Makefile
index b262c0a..759bc43 100644
--- a/Makefile
+++ b/Makefile
@@ -145,9 +145,13 @@ UPROGS=\
 	$U/_logstress\
 	$U/_forphan\
 	$U/_dorphan\
+	$U/_p1\
+	$U/_p2\
+	$U/_p3\
+	$U/_p4\
 
-fs.img: mkfs/mkfs README $(UPROGS)
-	mkfs/mkfs fs.img README $(UPROGS)
+fs.img: mkfs/mkfs README sample.txt $(UPROGS)
+	mkfs/mkfs fs.img README sample.txt $(UPROGS)
 
 -include kernel/*.d user/*.d
 
diff --git a/kernel/syscall.c b/kernel/syscall.c
index 076d965..45aa501 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -101,6 +101,8 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_knockknock(void);
+extern uint64 sys_getProcessStates(void);
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,6 +128,8 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_knockknock] sys_knockknock,
+[SYS_getProcessStates] sys_getProcessStates,
 };
 
 void
diff --git a/kernel/syscall.h b/kernel/syscall.h
index 3dd926d..32f3558 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,5 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_knockknock 22
+#define SYS_getProcessStates 23
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 419e727..7dcd936 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -25,7 +25,9 @@ sys_getpid(void)
 uint64
 sys_fork(void)
 {
-  return kfork();
+  int f=kfork();
+  printf("fork: created child with pid: %d\n",f);
+  return f;
 }
 
 uint64
@@ -107,3 +109,31 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+void
+sys_knockknock(void)
+{
+  int p=myproc()->pid;
+  printf("I am %d, Who's there?\n",p);
+  return;
+}
+
+void
+sys_getProcessStates(void)
+{
+  static char *states[] = {
+    [UNUSED]    "unused",
+    [USED]      "used",
+    [SLEEPING]  "sleep ",
+    [RUNNABLE]  "runble",
+    [RUNNING]   "run   ",
+    [ZOMBIE]    "zombie"
+  };
+  struct proc *p;
+  for(p = myproc(); p < &myproc()[NPROC]; p++) {
+    if(p->state != UNUSED) {
+      printf("Process %d was found in %s state\n",p->pid,states[p->state]);
+    }
+  }
+  return;
+}
\ No newline at end of file
diff --git a/sample.txt b/sample.txt
new file mode 100644
index 0000000..c39eda7
--- /dev/null
+++ b/sample.txt
@@ -0,0 +1,2 @@
+olaa
+elsa
diff --git a/user/p1.c b/user/p1.c
new file mode 100644
index 0000000..46897c2
--- /dev/null
+++ b/user/p1.c
@@ -0,0 +1,27 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char* argv[]) {
+    if (argc != 2)
+        printf("Usage: p1 <filename>\n");
+    else
+        printf("%s\n", argv[1]);
+
+    // Write your code here
+    
+    int f=fork();
+    if(f==0){
+        printf("in child %d\n",getpid());
+        close(0);
+        open(argv[1],0);
+        char* arr[2]={"cat",'\0'};
+        exec("cat",arr);
+    }
+    else if(f>0){
+        wait(0);
+        printf("%d,child: %d\n",getpid(),f);
+    }
+
+    exit(0);
+}
diff --git a/user/p2.c b/user/p2.c
new file mode 100644
index 0000000..a24eb65
--- /dev/null
+++ b/user/p2.c
@@ -0,0 +1,19 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char* argv[]) {
+
+    // Write your code here
+    
+    int f=fork();
+    if(f==0){
+        printf("in child %d\n",getpid());
+    }
+    else if(f>0){
+        wait(0);
+        printf("%d,child: %d\n",getpid(),f);
+    }
+
+    exit(0);
+}
diff --git a/user/p3.c b/user/p3.c
new file mode 100644
index 0000000..f1d7592
--- /dev/null
+++ b/user/p3.c
@@ -0,0 +1,7 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+int main() {
+    knockknock();
+    exit(0);
+}
diff --git a/user/p4.c b/user/p4.c
new file mode 100644
index 0000000..2751fca
--- /dev/null
+++ b/user/p4.c
@@ -0,0 +1,31 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main() {
+    int i, pid;
+
+    printf("Parent pid: %d\n", getpid());
+
+    for (i = 0; i < 3; i++) {
+        pid = fork();
+        if (pid == 0) {
+            pause(10);
+            exit(0);
+        }
+    }
+
+    pause(5);
+    getProcessStates();
+    pause(30);
+
+    printf("\n=== Process states after exiting ===\n");
+    getProcessStates();
+
+    for (i = 0; i < 3; i++) wait(0);
+
+    printf("\n=== Process states reaping ===\n");
+    getProcessStates();
+
+    exit(0);
+}
diff --git a/user/user.h b/user/user.h
index ac84de9..314fc50 100644
--- a/user/user.h
+++ b/user/user.h
@@ -24,6 +24,8 @@ int getpid(void);
 char* sys_sbrk(int,int);
 int pause(int);
 int uptime(void);
+void knockknock(void);
+void getProcessStates(void);
 
 // ulib.c
 int stat(const char*, struct stat*);
diff --git a/user/usys.pl b/user/usys.pl
index c5d4c3a..a645f61 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -42,3 +42,5 @@ entry("getpid");
 entry("sbrk");
 entry("pause");
 entry("uptime");
+entry("knockknock");
+entry("getProcessStates");
\ No newline at end of file
